### PR TITLE
SQLite options for Log::WRITER_SQL trading data safety for performance

### LINKED ISSUE #3987: SQLite Log writer painfully slow.
The SQLite log writer is painfully slow when used on a standard FS.

From my tests it performs around x10 slower than the normal asciiwriter, running on an SSD on Ubuntu 24.04.

This is likely caused by the synchronous and journal flags being default.

Suggestion:
Adding configurable flags for the SQLite log writer, that works the same like Broker::SQLiteOptions

### PR DESCRIPTION
**Pull request by Mymaqn**
The idea of this commit is based off of issue #2698 which allowed the broker data stores to change the SQLite pragmas `synchronous` and `journal_mode` to increase the performance of writes at the cost of data safety.

I was very sad to see, as someone who has a real use case for the Log::WRITER_SQL on persistent disks, that it was not an option to toggle these pragmas as well for the SQL log writer

This commit aims to fix that, adding options in the form of LogSQLite::synchronous and LogSQLite::journal_mode which can allow for faster writes on non-ramdisk filesystems at the cost of some data integrity.

This commit should close discussion #1870 and close #3987 in a relatively okay way.

This is my first commit every to a larger repository with many contributors, so I'm sure that I may have made some mistakes doing this PR. Please inform me if I need to change anything.

P.S. I was struggling with the pre-commit log formatter, creating inconsistencies in my switch cases, but decided to leave it in so the clang-format tests pass. But I don't like how it looks, and I don't know if anyone else agrees.

Thank you in advance for your patience and help

### CODE REVIEW

**Review by 0xxon — COMMENTED** [review]
This looks mostly fine to me - I just think that there should be an option to not execute any or the pragmas.

Does this make sense to you?

It also seems like it would be nice to have this setting changeable on a per-database-basis. However, that does add a bunch more work, and this is already better than our current status quo.

File: scripts/base/frameworks/logging/writers/sqlite.zeek [hunk 1]
```diff
@@ -21,5 +21,38 @@ export {
...
+	type SQLiteJournalMode: enum {
+		SQLITE_JOURNAL_MODE_DELETE,
+		SQLITE_JOURNAL_MODE_TRUNCATE,
+		SQLITE_JOURNAL_MODE_PERIST,
+		SQLITE_JOURNAL_MODE_MEMORY,
+		SQLITE_JOURNAL_MODE_WAL,
+		SQLITE_JOURNAL_MODE_OFF,
+	};
+	
+	## If set, runs the PRAGMA synchronous statement with the
```
[the remark this answers is not in the document]
Comment (Mymaqn) [reply scripts/base/frameworks/logging/writers/sqlite.zeek:42]: I like the first idea of using an extra enum value for both, to not execute the pragma, which it is automatically set to. I'll get to working on this right away

[the remark this answers is not in the document]
Comment (Mymaqn) [reply scripts/base/frameworks/logging/writers/sqlite.zeek]: Having the settings changeable on a per-database-basis would be really nice as well. But yes that is a bunch more work. I think that would have to be implemented in the filters, which I haven't had a chance to take a look at yet.

**Comment by 0xxon** [comment]
Ok, that looks better to me, thank you very much.

Would it perhaps be possible to add a testcase that uses these new options, similar to the sqlite writer testcases that already exist in `testing/btest/scripts/base/frameworks/logging`? :)

**Comment by Mymaqn** [comment]
> Ok, that looks better to me, thank you very much.
> 
> Would it perhaps be possible to add a testcase that uses these new options, similar to the sqlite writer testcases that already exist in `testing/btest/scripts/base/frameworks/logging`? :)

I would love to, but I also have to ask for help here.
I've tried to figure out how the testing stuff works, but can't really seem to grasp it. 
Do you guys have some documentation somewhere on how to work with and create new tests? 

I'm guessing we would just need something that tests whether the pragma's get set properly, but the only way this can be checked, is if we use the same DB connection as is used for writing (only journal mode WAL is persistent between runs), and I'm not sure how I can inject myself or receive that connection. Do you have any ideas?

**Comment by 0xxon** [comment]
Apologies for the slow reply.

I think a test just using the options to make sure that they don't break anything (so raise any sqlite errors that cause the logger to fail) would be sufficient.

You are right that it would be hard to test that the options take effect. That seems fine to me in this circumstance.

**Comment by 0xxon** [comment]
Superseded by #4063

**Pull request closed without merging**